In [108]:
import numpy as np

Задача 1

Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

In [109]:
def combinations(n: int, k: int) -> int:
    """Число сочетаний.
    """
    
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

Вероятность ответить на вопрос правильно 1/4, количество разных событий будет 6, от 0 правильных ответов до 5. Формула Бернулли:$$P(X=k)=C_n^k \cdot p^k \cdot (1-p)^{n-k}$$ p = 1/4, n = 5, k = \[0,...5]

In [110]:
p = 0.25
n = 5
probability = np.array([combinations(n,k) * p**k * (1 - p)**(n-k) for k in range(n+1)])
k = np.arange(n+1)
expected_value = (probability * k).sum()
dispersion = (probability * (k - expected_value)**2).sum()
print("Вероятсностьи получить значения от 0 до 5 равны соответственно:", probability)
print("Математическое ожидание:", expected_value)
print("Дисперсия:", dispersion)


Вероятсностьи получить значения от 0 до 5 равны соответственно: [0.23730469 0.39550781 0.26367188 0.08789062 0.01464844 0.00097656]
Математическое ожидание: 1.25
Дисперсия: 0.9375


Либо простым способом: $$M(X) = np$$ $$D(X)=np(1-p)$$<br>
$M(X)=5*0.25=1.25$<br>
$D(X)=1.25*(1-0.25)=0.9375$

Задача 2

Пользователь получает в среднем 10 писем со спамом на свой почтовый ящик за сутки. Найти число N, такое, что с вероятностью 0.95 пользователь получит не более N писем со спамом за текущий день.


Распределение пуассона: $$P_m = \frac{\lambda^k}{k!}e^{-\lambda}$$<br>
Будем складывать вероятности получить письма от 0 до некого N и прекратим когда суммарная вероятность превысит 0.95.

In [111]:
def puasson(lamb,k):
    """возвращает вероятность по формуле Пуассона, для лямбды и к.
    """
    return (lamb ** k) * (np.math.e ** (-lamb)) / np.math.factorial(k)

counter = -1
s = 0
lamb = 10
while s < 0.95:
    counter += 1
    s += puasson(lamb, counter)    
print(f"с вероятностью {s} пользователь получит не более {counter} писем со спамом за текущий день")



с вероятностью 0.9512595966960216 пользователь получит не более 15 писем со спамом за текущий день


Задача 3

Производятся выстрелы по мишени. Вероятность попадания в мишень при одном выстреле равна 0.01. Сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 6 раз будет совершено попадание?

Подсказка: 1) "Вероятность попасть k раз при n выстрелах" - на какое распределение это похоже? 2) А если нам нужна вероятность P(X >= k), а не P(X = k)? 3) Здесь предстоит немножко покодить.

тут мы увеличиваем количество выстрелов и считаем сумму вероятностей попвдания больше 6 выстрелов. когда она становится больше 0.9 мы прекращаем. это крайне долгий процесс, во всяком случае в таком исполнении.

In [112]:
p = 0.01
n = 0
k = 6
s = 0
while s < 0.9:
    n += 1 
    probability = np.array([combinations(n,k) * p**k * (1-p)**(n-k) for k in range(n+1)])
    s = sum(probability[6:])
print(s, n)

0.9003627121531275 926


Тут более быстрое исполнение, мы считаем что сумму вероятностей попасть от 0 до 5 выстрелов за n испытаний. она равна единица минус вероятность того что мы считаем в первом варианте. Ну а так как тут всего считать сумму 6 формул Бернулли. то получается гораздо быстрее при том же количестве циклов. скорее всего должно быть ~ в n раз быстрее чем в первом варианте(но это не точно).

In [114]:
p = 0.01
n = 5
k = 6
s = 1 # тут суммарная вероятность уменьшается с 1 так как промазать от 1 до 6 ракрайне вероятно.
while s > 0.1: #0.1 это 1 - 0.9
    n += 1 
    probability = np.array([combinations(n,k) * p**k * (1-p)**(n-k) for k in range(6)])
    s = sum(probability)
print(s, n)

0.09963728784686457 926


Плучается надо 926 выстрелов. 